src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1510,76 +1564,47 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 	List *workerList = NIL;
 	bool shardsPresent = false;
 	uint64 shardId = INVALID_SHARD_ID;
+	CmdType commandType = originalQuery->commandType;
+	bool localMultiShardModifyQuery = false;
 
 	*placementList = NIL;
 	prunedRelationShardList = TargetShardIntervalsForRouter(originalQuery,
 															restrictionContext,
 															&multiShardQuery);
 
-	/*
-	 * If multiShardQuery is true then it means a relation has more
-	 * than one shard left after pruning.
-	 */
 	if (multiShardQuery)
 	{
-		StringInfo errorMessage = makeStringInfo();
-		StringInfo errorHint = makeStringInfo();
-		CmdType commandType = originalQuery->commandType;
-		const char *commandName = "SELECT";
-
-		if (commandType == CMD_UPDATE)
-		{
-			commandName = "UPDATE";
-		}
-		else if (commandType == CMD_DELETE)
+		/*
+		 * If multiShardQuery is true and it is a type of SELECT query, then
+		 * return deferred error. We do not support multi-shard SELECT queries
+		 * with this code path.
+		 */
+		if (commandType == CMD_SELECT)
 		{
-			commandName = "DELETE";
+			prunedRelationShardList = NIL;
+			planningError = DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+										  NULL, NULL, NULL);
+			return planningError;
 		}
-
-		if (commandType == CMD_UPDATE || commandType == CMD_DELETE)
+		else if (commandType == CMD_UPDATE || commandType == CMD_DELETE)
 		{
-			List *rangeTableList = NIL;
-			RangeTblEntry *updateOrDeleteRTE = NULL;
-			DistTableCacheEntry *updateOrDeleteTableCacheEntry = NULL;
-			char *partitionKeyString = NULL;
-			char *partitionColumnName = NULL;
+			planningError = MultiShardModifyQuerySupported(originalQuery,
+														   prunedRelationShardList);
 
-			/* extract range table entries */
-			ExtractRangeTableEntryWalker((Node *) originalQuery, &rangeTableList);
-
-			updateOrDeleteRTE = GetUpdateOrDeleteRTE(rangeTableList);
-			updateOrDeleteTableCacheEntry =
-				DistributedTableCacheEntry(updateOrDeleteRTE->relid);
-
-			partitionKeyString = updateOrDeleteTableCacheEntry->partitionKeyString;
-			partitionColumnName = ColumnNameToColumn(updateOrDeleteRTE->relid,
-													 partitionKeyString);
+			if (planningError != NULL)
+			{
+				prunedRelationShardList = NIL;
+				return planningError;
+			}
 
-			appendStringInfo(errorHint, "Consider using an equality filter on "
-										"partition column \"%s\" to target a "
-										"single shard. If you'd like to run a "
-										"multi-shard operation, use "
-										"master_modify_multiple_shards().",
-							 partitionColumnName);
+			localMultiShardModifyQuery = true;
```

Comment:
This function gets a little messy due to changes that are specific to multi-shard update/delete below. Do you think it'd be possible to split up this function and call the relevant part, e.g.

```
if (multiShardQuery)
{
    if (commandType == CMD_SELECT)
    {
        ...
        return planningError;
    }

    PlanMultiShardModifyQuery(...);
}
else
{
    PlanSingleShardQuery(...);
}
```